### PR TITLE
Add "tslint:all" configuration

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [X] Addresses an existing issue: #2020
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update

#### Overview of change:

Minimal-change version of #2241

#### CHANGELOG.md entry:

[api] `tslint:all` configuration

### CODE REVIEW

File: src/configs/all.ts [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    "no-floating-promises": true,
+    "no-for-in-array": true,
+    "no-inferred-empty-object-type": true,
+    // "no-invalid-this": Won't this be deprecated?
+    "no-misused-new": true,
+    "no-null-keyword": true,
+    "no-shadowed-variable": true,
+    "no-string-literal": true,
+    "no-string-throw": true,
+    // "no-switch-case-fall-through": Won't this be deprecated?
```
Comment (adidahiya) [inline src/configs/all.ts]: yep, just delete this line

File: src/configs/all.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    "no-unbound-method": true,
+    "no-unsafe-any": true,
+    "no-unsafe-finally": true,
+    "no-unused-expression": true,
+    // "no-unused-variable": Won't this be deprecated?
```
Comment (adidahiya) [inline src/configs/all.ts]: still keeping it around; you probably ought to enable it here actually, given the kind of semantics you're going for for `tslint:all`

File: src/configs/all.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    "no-use-before-declare": true,
+    "no-var-keyword": true,
+    "no-void-expression": true,
+    "radix": true,
+    "restrict-plus-operands": true,
+    "strict-boolean-expressions": true,
+    "strict-type-predicates": true,
+    // "switch-default": Won't this be deprecated?
+    "triple-equals": true,
+    // "typeof-compare": Won't this be deprecated?
```
Comment (adidahiya) [inline src/configs/all.ts]: yep, remove

File: src/configs/all.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    "no-unused-expression": true,
+    // "no-unused-variable": Won't this be deprecated?
+    "no-use-before-declare": true,
+    "no-var-keyword": true,
+    "no-void-expression": true,
+    "radix": true,
+    "restrict-plus-operands": true,
+    "strict-boolean-expressions": true,
+    "strict-type-predicates": true,
+    // "switch-default": Won't this be deprecated?
```
Comment (adidahiya) [inline src/configs/all.ts]: we're going to keep it around actually, let's enable it here

File: test/ruleLoaderTests.ts [hunk 5]
```diff
@@ -99,28 +101,62 @@ describe("Rule Loader", () => {
...
         }
 
         assert.isFalse(testFailed, "List of rules doesn't match list of tests");
     });
+
+    it("includes every rule in 'tslint:all'", () => {
+        const expectedAllRules = everyRule().filter((ruleName) => {
+            // Some rules intentionally excluded
+            switch (ruleName) {
+                case "ban":
```
Comment (adidahiya) [inline test/ruleLoaderTests.ts]: I think you should maintain this blacklist in `configs/all.ts` as an array, and do `RULES_EXCLUDED_FROM_ALL_CONFIG.indexOf(ruleName) === -1` in this filter

File: test/ruleLoaderTests.ts [hunk 6]
```diff
@@ -99,28 +101,62 @@ describe("Rule Loader", () => {
...
+    });
 });
+
+function diffLists(actual: string[], expected: string[]): diff.IDiffResult[] {
+    return diff.diffLines(actual.join("\n"), expected.join("\n"));
+}
+
+function everyRule(): string[] {
+    return fs.readdirSync(srcRulesDir)
+            .filter((file) => /Rule.ts$/.test(file))
```
Comment (adidahiya) [inline test/ruleLoaderTests.ts]: nit: expected 8 spaces of indentation

**Comment by tonghae** [comment]
Is there a command to list all rules used in tslint.json including specified via "extends" property?

**Comment by ajafff** [comment]
@tonghae that's possible, but not with TSLint itself. I wrote a [linter](https://github.com/fimbullinter/wotan/blob/master/packages/wotan/README.md) that allows to [execute TSLint rules](https://github.com/fimbullinter/wotan/blob/master/packages/valtyr/README.md). It also knows how to handle the configuration correctly.

```sh
yarn add @fimbul/wotan @fimbul/valtyr
# or
npm install @fimbul/wotan @fimbul/valtyr
```

Note that you might need to use `./node_modules/.bin/wotan` instead of `wotan` in the following commands if you haven't modified your PATH or installed both packages globally.

```sh
# lookup the tslint.json for 'path/to/file.ts' like TSLint does
wotan show -m @fimbul/valtyr path/to/file.ts
# use a specific config file, not necessarily the one that TSLint's lookup finds
wotan show -m @fimbul/valtyr path/to/file.ts -c path/to/config.json
# you can even use config presets like 'tslint:all'
wotan show -m @fimbul/valtyr path/to/file.ts -c tslint:all
# output as JSON instead of YAML
wotan show -m @fimbul/valtyr path/to/file.ts -f json 
```

**Pull request merged**